File: classic-puzzle-easy/go/power-of-thor-episode-1.go
package main

import "fmt"

type Thor struct {
	GoTo string
	X    int
	Y    int
}

func main() {
    // lightX: the X position of the light of power
    // lightY: the Y position of the light of power
    // initialTx: Thor's starting X position
    // initialTy: Thor's starting Y position
    var lightX, lightY, initialTx, initialTy int
    fmt.Scan(&lightX, &lightY, &initialTx, &initialTy)
    thor := Thor{X: initialTx, Y: initialTy}
    for {
        // remainingTurns: The remaining amount of turns Thor can move. Do not remove this line.
        var remainingTurns int
        fmt.Scan(&remainingTurns)
        
        
		walkToLight(&thor, lightX, lightY) // pointer semantics are used to mutate the value of a struct. So the walkToLight function accepts a pointer to a Thor object.
		fmt.Println(thor.GoTo)
    }
}

func walkToLight(thor *Thor, lightX int, lightY int) {
	switch {
		case thor.X < lightX && thor.Y == lightY:
			thor.GoTo = "E"
			thor.X++
		case thor.X > lightX && thor.Y == lightY:
			thor.GoTo = "W"
			thor.X--
		case thor.X == lightX && thor.Y < lightY:
			thor.GoTo = "S"
			thor.Y++
		case thor.X > lightX && thor.Y < lightY:
			thor.GoTo = "SW"
			thor.X--
			thor.Y++
		case thor.X < lightX && thor.Y < lightY:
			thor.GoTo = "SE"
			thor.X++
			thor.Y++
		case thor.X == lightX && thor.Y > lightY:
			thor.GoTo = "N"
			thor.Y--
		case thor.X < lightX && thor.Y > lightY:
			thor.GoTo = "NE"
			thor.X++
			thor.Y--
		case thor.X > lightX && thor.Y > lightY:
			thor.GoTo = "NW"
			thor.X--
			thor.Y--
		default:
			thor.GoTo = " "
	}
}

File: classic-puzzle-easy/csharp/power-of-thor-episode-1.cs
using System;
using System.Linq;
using System.IO;
using System.Text;
using System.Collections;
using System.Collections.Generic;

class Player
{
    static void Main(string[] args)
    {
        string[] inputs = Console.ReadLine().Split(' ');
        int lightX = int.Parse(inputs[0]); // the X position of the light of power
        int lightY = int.Parse(inputs[1]); // the Y position of the light of power
        int initialTx = int.Parse(inputs[2]); // Thor's starting X position
        int initialTy = int.Parse(inputs[3]); // Thor's starting Y position

        var actualTPosition = new Tuple<int, int>(initialTx, initialTy);

        // game loop
        while (true)
        {
            int actualTx = int.Parse(inputs[2]); // Thor's X position
            int actualTY = int.Parse(inputs[3]); // Thor's Y position
            int remainingTurns = int.Parse(Console.ReadLine()); // The remaining amount of turns Thor can move. Do not remove this line.

            actualTPosition = WalkToLight(actualTPosition, lightX, lightY);
        }
    }

    static Tuple<int, int> WalkToLight(Tuple<int, int> actualPosition, int lightX, int lightY){
            if (actualPosition.Item2 == lightY && actualPosition.Item1 < lightX) {
                Console.WriteLine("E");
                return new Tuple<int, int>(actualPosition.Item1 + 1, actualPosition.Item2);
            } else if (actualPosition.Item2 == lightY && actualPosition.Item1 > lightX) {
                Console.WriteLine("W");
                return new Tuple<int, int>(actualPosition.Item1 - 1, actualPosition.Item2);
            } else if (actualPosition.Item1 == lightX && actualPosition.Item2 < lightY) {
                Console.WriteLine("S");
                return new Tuple<int, int>(actualPosition.Item1, actualPosition.Item2 + 1);
            } else if (actualPosition.Item1 > lightX && actualPosition.Item2 < lightY) {
                Console.WriteLine("SW");
                return new Tuple<int, int>(actualPosition.Item1 - 1, actualPosition.Item2 + 1);
            } else if (actualPosition.Item1 < lightX && actualPosition.Item2 < lightY) {
                Console.WriteLine("SE");
                return new Tuple<int, int>(actualPosition.Item1 + 1, actualPosition.Item2 + 1);
            } else if (actualPosition.Item1 == lightX && actualPosition.Item2 > lightY) {
                Console.WriteLine("N");
                return new Tuple<int, int>(actualPosition.Item1, actualPosition.Item2 - 1);
            } else if (actualPosition.Item1 < lightX && actualPosition.Item2 > lightY) {
                Console.WriteLine("NE");
                return new Tuple<int, int>(actualPosition.Item1 - 1, actualPosition.Item2 + 1);
            } else if (actualPosition.Item1 > lightX && actualPosition.Item2 > lightY) {
                Console.WriteLine("NW");
                return new Tuple<int, int>(actualPosition.Item1 + 1, actualPosition.Item2 + 1);
            }

            return new Tuple<int, int>(actualPosition.Item1, actualPosition.Item2);
    }
}